### PR TITLE
storage: add option to poison intents on EndTransaction

### LINKED ISSUE #18901: tpcc loadgen produces foreign key violation
@jordanlewis this happened to me once. Does it look like a server problem or a client problem?

```
[andrei:~/work/src/github.com/cockroachdb/loadgen/tpcc] master 49s 1 ± ./tpcc -concurrency=128
_time______opName__ops/s(inst)__ops/s(cum)__p50(ms)__p95(ms)__p99(ms)_pMax(ms)
   1s        tpmC         30.0        29.8    637.5    973.1   1006.6   1006.6
   2s        tpmC         34.0        31.9   1208.0   1744.8   1879.0   1879.0
2017/09/28 16:46:22 error in tpmC: pq: foreign key violation: value [0 6 3162] not found in order@primary [o_w_id o_d_id o_id]
```

### PR DESCRIPTION
**Pull request by spencerkimball**
This allows `EndTransaction` to be called out of band from the normal
client workflow in such a way that the transaction can be aborted but
will nevertheless leave an abort span entry to prevent a still-active
transaction from reading inconsistent intra-transaction state.

Fixes #18901

### CODE REVIEW

**Comment by bdarnell** [comment]
I like this better than #24048, but it needs a test.

---

Reviewed 1 of 8 files at r1.
Review status: 1 of 8 files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[pkg/storage/replica_command.go, line 300 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/24074#-L83kzRwFc4fYlE4MF6l:-L83kzRwFc4fYlE4MF6m:blerz4r) ([raw file](https://github.com/cockroachdb/cockroach/blob/76599b123562c92b312f06f93c404f24de9931be/pkg/storage/replica_command.go#L300)):*
> ```Go
> 			// Use alwaysReturn==true because the transaction is definitely
> 			// aborted, no matter what happens to this command.
> 			return result.FromEndTxn(reply.Txn, true /* alwaysReturn */, false /* poison */), nil
> ```

Why does this line and the next hard-code false instead of `args.Poison`? It shouldn't get here because it would require the failed heartbeat to come after a regular abort, but I think it would still be more logically consistent to use args.Poison here. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/24074#-:-L83lHQQYpI9sZjae0by:b-ooph05)*

**Comment by spencerkimball** [comment]
Added a test.

---

Review status: 1 of 8 files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[pkg/storage/replica_command.go, line 300 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/24074#-L83kzRwFc4fYlE4MF6l:-L83wJzvS0tmbI4obkxA:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/76599b123562c92b312f06f93c404f24de9931be/pkg/storage/replica_command.go#L300)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Why does this line and the next hard-code false instead of `args.Poison`? It shouldn't get here because it would require the failed heartbeat to come after a regular abort, but I think it would still be more logically consistent to use args.Poison here. 
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/24074#-:-L83wHxJbBBxvZd1vX5P:b-7gclag)*

File: pkg/kv/txn_coord_sender.go [hunk 1]
```diff
@@ -1010,13 +1010,18 @@ func (tc *TxnCoordSender) tryAsyncAbort(ctx context.Context) {
 		tc.AnnotateCtx(context.Background()), "kv.TxnCoordSender: aborting txn", func(ctx context.Context) {
 			// Use the wrapped sender since the normal Sender does not allow
 			// clients to specify intents.
+			log.Infof(ctx, "sending end transaction request to abort")
```
Comment (andreimatei) [inline pkg/kv/txn_coord_sender.go]: `VEventf` or if you ever want this in the shared log, pls add more info - the txn id or something

Comment (spencerkimball) [reply pkg/kv/txn_coord_sender.go]: Had some leftover debug statements. Removed.

**Comment by bdarnell** [comment]
:lgtm:

---

Reviewed 4 of 8 files at r1, 1 of 4 files at r2, 3 of 3 files at r3.
Review status: all files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[pkg/kv/txn_coord_sender_test.go, line 728 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/24074#-L87zEU_0y488UDl3vyl:-L87zEU_0y488UDl3vym:bh8erdp) ([raw file](https://github.com/cockroachdb/cockroach/blob/98f6e4de66649ddfa4dc66c2335d6c2b07348fb4/pkg/kv/txn_coord_sender_test.go#L728)):*
> ```Go
> 	// Add a testing request filter which pauses a get request for the
> 	// key until after the signal channel is closed.
> 	key := roachpb.Key("a")
> ```

Intents are handled differently depending on whether or not they are on the same range as the transaction record. We should test two different keys here: the transaction's anchor key, and one on a different range. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/24074#-:-L88-FTLbzxBbko3Mcy0:bnfp4nl)*

**Comment by spencerkimball** [comment]
Review status: 9 of 10 files reviewed at latest revision, 2 unresolved discussions.

---

*[pkg/kv/txn_coord_sender_test.go, line 728 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/24074#-L87zEU_0y488UDl3vyl:-L88iA49qucOPsCmAXA6:b-cwb2hg) ([raw file](https://github.com/cockroachdb/cockroach/blob/98f6e4de66649ddfa4dc66c2335d6c2b07348fb4/pkg/kv/txn_coord_sender_test.go#L728)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Intents are handled differently depending on whether or not they are on the same range as the transaction record. We should test two different keys here: the transaction's anchor key, and one on a different range. 
</blockquote></details>

OK, new test added which checks behavior across ranges.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/24074)*

**Comment by bdarnell** [comment]
LGTM

---

Review status: 9 of 10 files reviewed at latest revision, 2 unresolved discussions, some commit checks pending.

---



*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24074#-:-L88j1GVDBjNWaztbNxF:bb6ywmx)*

**Pull request merged**